src/Tablebot/Plugins/Administration.hs

```diff
@@ -0,0 +1,180 @@
+import Control.Monad.Trans.Reader (ask)
+import Data.Text (Text, pack)
+import qualified Data.Text as T
+import Database.Persist (Entity, Filter, entityVal, (==.))
+import Discord (stopDiscord)
+import Discord.Types
+import Tablebot.Handler.Administration
+import Tablebot.Handler.Types (CompiledPlugin (compiledName))
+import Tablebot.Plugin
+import Tablebot.Plugin.Database
+import Tablebot.Plugin.Discord (sendMessage)
+import Tablebot.Plugin.Permission (requirePermission)
+import Tablebot.Plugin.SmartCommand
+import Text.RawString.QQ
+
+-- | @SS@ denotes the type returned by the command setup. Here its unused.
+type SS = [Text]
+
+blacklist :: Command SS
+blacklist = Command "blacklist" (parseComm blacklistComm)
+
+blacklistComm ::
+  WithError
+    "Unknown quote functionality."
+    ( Either
+        ( Either
+            (Exactly "add", String)
+            (Exactly "remove", String)
+        )
+        (Exactly "list")
+    ) ->
+  Message ->
+  DatabaseDiscord SS ()
+blacklistComm (WErr (Left (Left (_, pLabel)))) = addBlacklist pLabel
+blacklistComm (WErr (Left (Right (_, pLabel)))) = removeBlacklist pLabel
+blacklistComm (WErr (Right (_))) = listBlacklist
+
+addBlacklist :: String -> Message -> DatabaseDiscord SS ()
+addBlacklist pLabel m = requirePermission Superuser m $ do
+  known <- ask
+  -- It's not an error to add an unknown plugin (so that you can pre-disable a plugin you know you're about to add),
+  -- but emmit a warning so people know if it wasn't deliberate
+  when ((pack pLabel) `notElem` known) $ sendMessage m "Warning, unknown plugin"
+  extant <- exists [PluginBlacklistLabel ==. pLabel]
+  if not $ extant
+    then do
+      _ <- insert $ PluginBlacklist pLabel
+      sendMessage m "Plugin added to blacklist. Please reload for it to take effect"
+    else sendMessage m "Plugin already in blacklist"
+
+removeBlacklist :: String -> Message -> DatabaseDiscord SS ()
+removeBlacklist pLabel m = requirePermission Superuser m $ do
+  extant <- selectKeysList [PluginBlacklistLabel ==. pLabel] []
+  if not $ null extant
+    then do
+      _ <- delete (head extant)
+      sendMessage m "Plugin removed from blacklist. Please reload for it to take effect"
+    else sendMessage m "Plugin not in blacklist"
+
+-- | @listBlacklist@ shows a list of the plugins eligible for disablement (those not starting with _),
+--  along with their current status.
+listBlacklist :: Message -> DatabaseDiscord SS ()
+listBlacklist m = requirePermission Superuser m $ do
+  bl <- selectList allBlacklisted []
+  pl <- ask
+  sendMessage m (format pl (blacklisted bl))
+  where
+    allBlacklisted :: [Filter PluginBlacklist]
+    allBlacklisted = []
+    -- Oh gods, so much formatting.
+    format :: [Text] -> [Text] -> Text
+    format p bl = "**Plugins:**\n```\n" <> (T.concat $ map (format' len' bl) (filter (disableable . T.uncons) p)) <> "```" <> unknown (filter (`notElem` p) bl)
+      where
+        len' :: Int
+        len' = maximum $ map T.length p
```

Comment:
This isn't very clear to me - I get that it's providing a formatted version, but it isn't very clear how. (The names don't help a lot here - e.g. what does `format'` do?) It should at least be multilined (newline after some `<>`).

Reply:
Take a look at it now. I've hopefully cleaned it up a fair bit.